### PR TITLE
chore(s3): reduce false positive in s3 public check

### LINKED ISSUE #4257: [Bug]: False Positive on check s3_bucket_public_access when Conditions in Policy
### Steps to Reproduce

1. prowler aws -c s3_bucket_public_access
2. aws
3. Single Account
4. Failed finding: S3 Bucket sample-restricted-bucket has public access due to bucket policy.

### Expected behavior

S3 Bucket sample-restricted-bucket should have passed the check.

### Actual Result with Screenshots or Logs

Failed check: S3 Bucket sample-restricted-bucket has public access due to bucket policy.

### How did you install Prowler?

From pip package (pip install prowler)

### Environment Resource

7. Workstation (Linux Subsystem)

### OS used

5. Ubuntu 22.04 (WSL)

### Prowler version

Prowler 4.2.4 

### Pip version

pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)

### Context

The check only looks for Principal = "*" and Effect = "Allow" without investigating other contexts to the statement, such as Conditions or Resources.
https://github.com/prowler-cloud/prowler/blob/61b9ecc2149d98cb013967f381d15e159f36530d/prowler/providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py#L51

EX. For this policy, the check should pass as the condition restricts **public** access:
` {
            "Sid": "AllowVPCAccess",
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:*",
            "Resource": [
                "arn:aws:s3:::sample-restricted-bucket/*",
                "arn:aws:s3:::sample-restricted-bucket"
            ],
            "Condition": {
                "StringEquals": {
                    "aws:sourceVpc": "vpc-12345678abcdef"
                }
            }
        }`

### PR DESCRIPTION
**Pull request by puchy22**
### Context

Fixes #4257 

Not checking conditions in policy statements

### Description

Add new policy function to check policy conditions and improve checks using this functions. Everything added have been tested properly.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by MrCloudSec — DISMISSED** [review]
LGTM, love this fix!

**Comment by jmanduca-psfy** [comment]
Just tried it out, it works for some bucket policies but not others - this looks like it is due to the case of your `no_public_conditions`, as the value in policy is not case-sensitive, causing some that were manually created to fail. 
Fail: 
`            "Condition": {
                "StringEquals": {
                    "aws:sourcevpc": "vpc-12334"`
Pass:
`            "Condition": {
                "StringEquals": {
                    "aws:SourceVpc": "vpc-12334"`

**Comment by jfagoagas** [comment]
> Just tried it out, it works for some bucket policies but not others - this looks like it is due to the case of your `no_public_conditions`, as the value in policy is not case-sensitive, causing some that were manually created to fail. Fail: ` "Condition": { "StringEquals": { "aws:sourcevpc": "vpc-12334"` Pass: ` "Condition": { "StringEquals": { "aws:SourceVpc": "vpc-12334"`

Great catch, we need to do the match that way, please @puchy22 change it and add some tests to cover that.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
See above comment and check the function `is_condition_block_restrictive` at https://github.com/prowler-cloud/prowler/blob/9fbd627f9aec3edee4f7ddc25849faa11db6187e/prowler/providers/aws/lib/policy_condition_parser/policy_condition_parser.py

We should unify both functions since they are similar.

**Comment by MrCloudSec** [comment]
@puchy22 where did you get the `no_public_conditions` from?

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 1]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
                     ) and "Condition" not in statement:
                         return True
     return False
+
+
+def check_full_service_access(service: str, policy: dict) -> bool:
```
Comment (MrCloudSec) [inline prowler/providers/aws/services/iam/lib/policy.py]: Is this function being used?

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 2]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
...
+    if condition_statement.get("IpAddress", {}):
+        # We need to transform the condition_statement into lowercase
+        condition_statement["IpAddress"] = {
+            k.lower(): v for k, v in condition_statement["IpAddress"].items()
+        }
+
+        if condition_statement["IpAddress"].get("aws:sourceip", ""):
+            if not isinstance(condition_statement["IpAddress"]["aws:sourceip"], list):
+                condition_statement["IpAddress"]["aws:sourceip"] = [
+                    condition_statement["IpAddress"]["aws:sourceip"]
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/lib/policy.py]: Could you define the following as constants?
- `aws:sourceip`
- `IpAddress`

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 3]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
...
+    Keyword arguments:
+    condition_statement -- The policy condition to check. For example:
+        {
+            "IpAddress": {
+                "aws:SourceIp": "X.X.X.X"
+            }
+        }
+    """
+
+    is_from_private_ip = False
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/lib/policy.py]: Could you add a try/except block for the whole function?

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 4]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
...
+                condition_statement["IpAddress"]["aws:sourceip"] = [
+                    condition_statement["IpAddress"]["aws:sourceip"]
+                ]
+
+            for ip in condition_statement["IpAddress"]["aws:sourceip"]:
+                try:
+                    # Select if IP address or IP network searching in the string for '/'
+                    if "/" in ip:
+                        if not ipaddress.ip_network(ip, strict=False).is_private:
+                            break
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/lib/policy.py:113]: Cover this line with a test please.

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 5]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
...
+                    else:
+                        if not ipaddress.ip_address(ip).is_private:
+                            break
+
+                except ValueError:
+                    logger.error(f"Invalid IP: {ip}")
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/lib/policy.py]: Cover this line with a test please.

File: prowler/providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py [hunk 6]
```diff
@@ -46,26 +52,29 @@ def execute(self):
...
-                                            statement["Principal"]["AWS"], str
-                                        ):
-                                            principals = [statement["Principal"]["AWS"]]
-                                        else:
-                                            principals = statement["Principal"]["AWS"]
+                                    if "*" == statement["Principal"]:
+                                        report.status = "FAIL"
+                                        report.status_extended = f"S3 Bucket {bucket.name} has public access due to bucket policy."
+                                    elif "AWS" in statement["Principal"]:
+                                        principals = (
```
Comment (jfagoagas) [inline prowler/providers/aws/services/s3/s3_bucket_public_access/s3_bucket_public_access.py:76]: Cover this line with a test please.

File: prowler/providers/aws/services/iam/lib/policy.py [hunk 7]
```diff
@@ -68,3 +73,96 @@ def is_policy_public(policy: dict) -> bool:
...
+        {
+            "IpAddress": {
+                "aws:SourceIp": "X.X.X.X"
+            }
+        }
+    """
+
+    is_from_private_ip = False
+
+    if condition_statement.get("IpAddress", {}):
```
Comment (jfagoagas) [inline prowler/providers/aws/services/iam/lib/policy.py]: I will move this line into the function caller, makes little sense to call the function if the condition is not `IpAddress`.

File: tests/providers/aws/services/iam/lib/policy_test.py [hunk 8]
```diff
@@ -89,3 +91,119 @@ def test_is_policy_public(self):
...
+                    "Action": "s3:*",
+                    "Resource": "arn:aws:s3:::example_bucket",
+                }
+            ]
+        }
+
+        assert check_full_service_access("s3", policy1)
+        assert not check_full_service_access("s3", policy2)
+        assert not check_full_service_access("s3", policy3)
+        assert not check_full_service_access("s3", policy4)
```
Comment (jfagoagas) [inline tests/providers/aws/services/iam/lib/policy_test.py]: Could you please create one test case for each policy? It's simpler to work with the tests are failing.

**Review by jfagoagas — CHANGES_REQUESTED** [review]

**Review by jfagoagas — APPROVED** [review]
This PR is pure gold 🏅 Thanks for the effort!!

**Review by MrCloudSec — APPROVED** [review]
🤩

**Pull request merged**